packages/logging-controller/src/LoggingController.ts

```diff
@@ -0,0 +1,180 @@
+ * and MetaMask support agents helping those users. These logs are typed as
+ * generic.
+ */
+export type GenericLog = {
+  type: LogType.GenericLog;
+  data: any;
+};
+
+/**
+ * Union of all possible log data structures.
+ */
+export type Log = EthSignLog | GenericLog;
+
+/**
+ * LogEntry is the entry that will be added to the logging controller state.
+ * It consists of a entry key that must be on of the Log union types, and an
+ * additional id and timestamp.
+ */
+export type LogEntry = {
+  id: string;
+  timestamp: number;
+  log: Log;
+};
+
+/**
+ * Logging controller state
+ *
+ * @property logs - An object of logs indexed by their ids
+ */
+export type LoggingControllerState = {
+  logs: {
+    [id: string]: LogEntry;
+  };
+};
+
+const name = 'LoggingController';
+
+export type AddLog = {
+  type: `${typeof name}:add`;
+  handler: LoggingController['addLog'];
+};
+
+export type LoggingControllerMessenger = RestrictedControllerMessenger<
+  typeof name,
+  AddLog,
+  never,
+  never,
+  never
+>;
+
+const metadata = {
+  logs: { persist: true, anonymous: false },
+};
+
+const defaultState = {
+  logs: {},
+};
+
+/**
+ * Controller that manages a list of logs for signature requests.
+ */
+export class LoggingController extends BaseControllerV2<
+  typeof name,
+  LoggingControllerState,
+  LoggingControllerMessenger
+> {
+  /**
+   * Creates a LoggingController instance.
+   *
+   * @param options - Constructor options
+   * @param options.messenger - An instance of the ControllerMessenger
+   * @param options.state - Initial state to set on this controller.
+   */
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: LoggingControllerMessenger;
+    state?: Partial<LoggingControllerState>;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: {
+        ...defaultState,
+        ...state,
+      },
+    });
+
+    this.messagingSystem.registerActionHandler(
+      `${name}:add` as const,
+      (log: Log) => this.addLog(log),
+    );
+  }
+
+  /**
+   * Recursive method to ensure no collisions of ids.
+   *
+   * We may want to end up using a hashing mechanism to make ids deterministic
+   * by the *data* passed in, and then make each key an array of logs that
+   * match that id.
+   *
+   * @returns unique id
+   */
+  generateId(): string {
+    const id = random();
+    if (this.state.logs[id]) {
+      return this.generateId();
+    }
+    return id;
+  }
+
+  /**
+   * Add log to the state.
+   *
+   * @param log - Log to add to the controller
+   */
+  addLog(log: Log) {
+    const newLog: LogEntry = {
+      id: this.generateId(),
+      timestamp: Date.now(),
+      log,
+    };
+
+    this.update((state) => {
+      state.logs[newLog.id] = newLog;
+    });
+  }
+
+  /**
+   * Removes all log entries.
+   */
+  clear() {
+    this.update((state) => {
+      state.logs = {};
+    });
+  }
+}
+
+export default LoggingController;
```

Comment:
Do we need to export this as the default, since we're already exported it as a named export?

Reply:
Update: It seems that we do this quite a lot in our controllers, including EnsController, so I can see where this came from. That said, I consider this an anti-pattern, as there's no real need for it (we don't expect developers to import individual files from packages, even though it's technically possible).